crates/kubelet/src/pod.rs

```diff
@@ -85,27 +86,88 @@ impl Pod {
         Some(self.annotations().get(key)?.as_str())
     }
 
-    /// Patch the pod status to update the phase.
-    pub async fn patch_status(&self, client: APIClient, phase: &Phase) {
-        let status = serde_json::json!(
+    /// Patch the pod status using the given status information.
+    pub async fn patch_status(&self, client: APIClient, status: Status) {
+        let name = self.name();
+        let api: Api<KubePod> = Api::namespaced(client, self.namespace());
+        let current_status = match api.get(name).await {
+            Ok(p) => match p.status {
+                Some(s) => s,
+                None => {
+                    error!("Pod is missing status information. This should not occur");
+                    return;
+                }
+            },
+            Err(e) => {
+                error!("Unable to fetch current status of pod {}, aborting status patch (will be retried on next status update): {:?}", name, e);
+                return;
+            }
+        };
+
+        // This section figures out what the current phase of the pod should be
+        // based on the container statuses
+        let current_statuses = status
+            .container_statuses
+            .into_iter()
+            .map(|s| s.to_kubernetes(name.to_owned()))
+            .collect::<Vec<KubeContainerStatus>>();
```

Comment:
Should we collect this into a `HashMap<String, KubeContainerStatus>` where the keys are the `name`s so we can do quick lookups in the filter below?